opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java

```diff
@@ -0,0 +1,475 @@
+          .record(callLatencyNanos / NANOS_PER_SECOND, attribute);
+    }
+  }
+
+  private static final class ServerTracer extends ServerStreamTracer {
+    @Nullable private static final AtomicIntegerFieldUpdater<ServerTracer> streamClosedUpdater;
+    @Nullable private static final AtomicLongFieldUpdater<ServerTracer> outboundWireSizeUpdater;
+    @Nullable private static final AtomicLongFieldUpdater<ServerTracer> inboundWireSizeUpdater;
+
+    /*
+     * When using Atomic*FieldUpdater, some Samsung Android 5.0.x devices encounter a bug in their
+     * JDK reflection API that triggers a NoSuchFieldException. When this occurs, we fall back to
+     * (potentially racy) direct updates of the volatile variables.
+     */
+    static {
+      AtomicIntegerFieldUpdater<ServerTracer> tmpStreamClosedUpdater;
+      AtomicLongFieldUpdater<ServerTracer> tmpOutboundWireSizeUpdater;
+      AtomicLongFieldUpdater<ServerTracer> tmpInboundWireSizeUpdater;
+      try {
+        tmpStreamClosedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(ServerTracer.class, "streamClosed");
+        tmpOutboundWireSizeUpdater =
+            AtomicLongFieldUpdater.newUpdater(ServerTracer.class, "outboundWireSize");
+        tmpInboundWireSizeUpdater =
+            AtomicLongFieldUpdater.newUpdater(ServerTracer.class, "inboundWireSize");
+      } catch (Throwable t) {
+        logger.log(Level.SEVERE, "Creating atomic field updaters failed", t);
+        tmpStreamClosedUpdater = null;
+        tmpOutboundWireSizeUpdater = null;
+        tmpInboundWireSizeUpdater = null;
+      }
+      streamClosedUpdater = tmpStreamClosedUpdater;
+      outboundWireSizeUpdater = tmpOutboundWireSizeUpdater;
+      inboundWireSizeUpdater = tmpInboundWireSizeUpdater;
+    }
+
+    private final OpenTelemetryMetricsModule module;
+    private final String fullMethodName;
+    private volatile boolean isGeneratedMethod;
+    private volatile int streamClosed;
+    private final Stopwatch stopwatch;
+    private volatile long outboundWireSize;
+    private volatile long inboundWireSize;
+
+    ServerTracer(OpenTelemetryMetricsModule module, String fullMethodName) {
+      this.module = checkNotNull(module, "module");
+      this.fullMethodName = fullMethodName;
+      this.stopwatch = module.stopwatchSupplier.get().start();
+    }
+
+    @Override
+    public void serverCallStarted(ServerCallInfo<?, ?> callInfo) {
+      // Only record method name as an attribute if isSampledToLocalTracing is set to true,
+      // which is true for all generated methods. Otherwise, programatically
+      // created methods result in high cardinality metrics.
+      isGeneratedMethod = callInfo.getMethodDescriptor().isSampledToLocalTracing();
+      io.opentelemetry.api.common.Attributes attribute =
+          io.opentelemetry.api.common.Attributes.of(
+              AttributeKey.stringKey(OpenTelemetryConstants.METHOD_KEY),
+              recordMethodName(fullMethodName, isGeneratedMethod));
```

Comment:
Read isGeneratedMethod from a local variable. That provides a small performance improvement (since it wouldn't need to be re-read), but more importantly it is generally a good idea to read volatiles into locals so their value doesn't change. Here it is fine, but it is a good habit.